### ISSUE TITLE
Default to W^X JIT on OpenBSD systems

### ISSUE THREAD

**Issue by hrydgard** [hrydgard/ppsspp#8943]
Looks like it's the way of the future:

https://marc.info/?l=openbsd-ports&m=147060153102724&w=2

Not sure if we actually support OpenBSD though :)

**Comment by unknownbrackets** [comment]
@jbeich should we do this on all BSDs?

We've recently implemented this for iOS and it appears to work, but it probably hasn't received a ton of thorough testing yet.

-[Unknown]

**Comment by jbeich** [comment]
Why not dogfood everywhere except maybe Windows?

FreeBSD doesn't support W^X yet unlike [some](http://blog.acumensecurity.net/fpt_wx_ext-1-a-rundown/) other BSDs . Testing b06359e snapshot with `PlatformIsWXExclusive()` -> `true` (see [port diff](https://gist.github.com/jbeich/e5759a98f572fdf8a4adffe2d351ba44)) showed no regression when playing Cave Story (from in-emulator Homebrew Store). I haven't tried to benchmark.

@lattera maybe able to test W^X works as intended on HardenedBSD.

**Comment by lattera** [comment]
If I'm given steps on how to properly test, I'd love to help. :)

**Comment by jbeich** [comment]
The test is mainly to catch crashes.
1. Check `glxinfo` supports OpenGL 2.0
2. Follow "port diff" link in my previous message (download as Raw)
3. Apply against emulators/ppsspp
4. Install the port (select SDL option to reduce number of dependencies)
5. Launch `ppsspp`
6. Select `Homebrew & Demos` tab
7. Select `Download from PPSSPP Homebrew Store`
8. Select `Cave Story` (or any other game)
9. Select `Install` then `Launch game`
10. Press Space on `English` and play... (Esc to invoke menu)
11. Check `procstat -v $(pgrep ppsspp) | awk 'NR == 1 || /wx/'`
12. Repeat starting from **4** step but remove files/patch-wxforce

**Comment by unknownbrackets** [comment]
Well, it can depend but there may be some impact to performance using W^X: we have to do more things.  And unlike a jit for say, JavaScript, we toggle protection more frequently, especially in games that modify their own code.

If `mprotect()` is universally cheap on BSD, and may potentially be needed, it's probably for the better to enable.

Note: the cost will most of the time only be during loading, not during runtime.  For example, it might affect how long it takes a scene to load, and initial stutter moving around in that scene.  Once the code is jitted (assuming it's not being modified frequently) performance should be the same.  Ideally, we don't want initial stutters, though, that's why I'm concerned about any perf impact.

-[Unknown]

**Comment by lattera** [comment]
(Responding from my phone via email, please excuse any brevity)

W^X, when implemented properly, won't allow mprotect calls between W and X
to succeed either. If a page has ever been marked as writable, it can never
be marked executable, even if the write bit is dropped in the future. And
vice-versa; if a page has ever been marked executable, it can never be
marked writable, even if the executable bit is dropped.

What you could do, is have a writable shadow page. You map the same pages
twice: once as RX and another time as W. Prior to execution, you drop the
writable shadow page. This is what Firefox now does with minimal
performance impact.

I'd be happy to help test and verify correct behavior. I'll take a look
later this week, if I have time.

**Comment by unknownbrackets** [comment]
Well, that definitely means more complexity.  PPSSPP does a lot of "prior to execution."  Do we need to drop and remap for each time, if we use a shadow?

PPSSPP compiles and then immediately executes small chunks of MIPS code (say, between 8 and 200 bytes of MIPS, typically) at a time.  In some cases, it will use "block linking" (an important optimization for tight loops) to modify a previously compiled basic block, and make it jump directly to another basic block (rather than back to dispatch to find the block.)

Just as an example, Dolphin does some of the same things, and also does more "backpatching" of certain memory read/write instructions (e.g. to handle TLB emulation, iirc.)  I think most console emulators jit by basic block.

The `PlatformIsWXExclusive` functionality currently just swaps pages between W and X, so it sounds like it will not work with a proper W^X implementation.

-[Unknown]

**Comment by hrydgard** [comment]
I'm not sure I agree to call such W^X rules "proper", more like "overzealous" - just disallowing pages that are both W and X at the same time should already make things harder for simple exploits. But maybe I'm missing something. But even if I do, it also seems to be that by allowing the shadow thing, you're right back where you started in terms of security?

**Comment by hrydgard** [comment]
https://jandemooij.nl/blog/2015/12/29/wx-jit-code-enabled-in-firefox/   seems to indicate that they actually do remap pages W->X and that the shadow page thing wasn't merged. Though that might be old information.

**Comment by lattera** [comment]
It looks like I was partially mistaken with Firefox. What Firefox does is:

```
mapping = mmap(NULL, PROT_READ | PROT_WRITE, MAP_ANON | MAP_SHARED, -1, 0);
write_all_the_javascript_things(mapping);
mapping = mmap(mapping, PROT_READ | PROT_EXEC, MAP_ANON | MAP_SHARED | MAP_FIXED, -1, 0);
```

By doing it that way, they get around W^X. They don't use shadow pages, but simply remap without using `mprotect` to upgrade from writable to executable. It appears that in userland, it's not even possible to create shadow pages. Attached is sample code that attempts to create a shadow page. You'll notice that the page simply gets remapped to RW on the second call to `mmap`.

So, that's probably the avenue that's best to take. First map with RW, then remap as RX. I'd stay far away from `mprotect` to modify page permissions, since PaX-based exploit mitigations will not be happy with you.

Sample code: [mem.c.txt](https://github.com/hrydgard/ppsspp/files/456777/mem.c.txt)

**Comment by hrydgard** [comment]
So um.. how is that any more secure than using mprotect? I don't get it, really, as it seems to have the same effect, just that you have to do it on the whole mapping instead of just a part of it.

That will probably not be very efficient for us.

**Comment by lattera** [comment]
I'm attaching a new version of the same sample code. It actually shows that calling `mmap` multiple times zeros out the pages. So my previous suggestion of remapping the pages as RX from RW actually won't work.

So, for OpenBSD, you could use `mprotect` to upgrade the pages from RW to RX, but you wouldn't be able to on HardenedBSD, since HardenedBSD uses the PaX model. Additionally, on HardenedBSD, you wouldn't be able to remap the pages as RX as remapping would zero out the pages.

If ppspp wen the `mprotect` route, we would simply have to add an exception to the ppsspp binary that says "disable `mprotect` restrictions." We do that for a number of applications already, so it wouldn't be difficult to to for ppsspp. I'd assume since this ticket is to enable W^X compat, we'd already have to disable pageexec restrictions, mprotect restrictions, or both.

The only other way around it would be to write the JITed code to a file, close and reopen the file as read only, and map the file into memory as RX. That would kill performance, though.

New sample code: [mem.c-rev02.txt](https://github.com/hrydgard/ppsspp/files/456889/mem.c-rev02.txt)

**Comment by hrydgard** [comment]
Right, I think we'll need that exception then. If we enable the recent semi-W^X iOS fixes we don't need RWX pages anymore at least, though we do need unlimited mprotect-switching between RW and RX.

**Comment by lattera** [comment]
Admittedly, it's an extremely hard problem to solve. Which is probably why OpenBSD went the route of lesser enforcement. PaX's model is generally extremely strict. But at least that lets us, as exploit mitigation authors, know which applications misbehave. ;)

If I get some extra time on Saturday or Sunday, I'll give your W^X compat work a try in HardenedBSD. I've got an extremely busy week this week and I'm headed to Serbia for EuroBSDcon next week.

**Comment by hrydgard** [comment]
Personally I feel that one of the charms of von Neumann machines is that data and instructions live in the same space and code can generate code, and these newfangled security restrictions take that flexibility away a little too much :) But I suppose for good reason in many cases..

**Comment by unknownbrackets** [comment]
Is this fixed with #10058?

-[Unknown]

**Comment by unknownbrackets** [comment]
I'm going to assume that this is fixed now, since it's been about 4.5 years without anyone saying it isn't.

-[Unknown]
